### ISSUE TITLE
Overall comments on the content of the repository

### ISSUE THREAD

**Issue by floridop** [SoapNoodles15/MNXB11-final-project#7]
This issue contains feedback about your project. In some cases it may 
contain actions to carry in order to get a final grade.

# Workplan

The workplan seem nice. There were no real dates, but if feels you
somewhat keep track of who does what when.
More precisely, Sebastian did, according to the git log.

# ChangeLog

The ChangeLog does not seem to follow the suggested standard (see end of manual)
and it's also missing the names of who did what. Entries where also
mostly updated by Sebastian.

# README.md

The document may have benefitted of a better use of MarkDown syntax,
that you could have learned by looking at the source code of all the 
README.md documents provided by the course and even the project template.

It is nevertheless quite detailed and I managed to build and execute your code.

# BASH Code - `cleaner_filter.sh`

The bash code contains interesting and complex regular expressions for grep. I suspect
these were partially inspired by ChatGPT since I did not teach any of this.

I am overall happy that you tried to reuse existing code, however, the use
of `awk` smells also like ChatGPT or StackOverflow. 
In general I do not appreciate an intensive use of `awk`, and I doubt 
any of you is an `awk` expert. 

I am referring to these lines:
https://github.com/SoapNoodles15/MNXB11-final-project/blob/6d5ff242b9bec082027f6dbb29047fd020a62657/cleaner_filter.sh#L78-L105

Since ColdestWarmest was mostly about comparing values, I would have 
appreciated much more that piece of code to be carried on using C++. 

I would not trust `awk` on numerical comparisons. 

# C++ Code

## main.cxx

As I already said during the exam, calling a BASH script from C++ should
be avoided at best. There are also security concerns about that: if 
I am an attacker I can change the external BASH script to delete all 
your files and your code will never know. 

You want your compiled stuff to do one thing and to do it the best,
without non-compiled dependencies at best. This is because you may 
want to use just the binary and not ship the binary together with 
scripts that might cause it to fail.

`while(true)` is also not a very good way to go. Modern interactive
coding is based on event-based  and instead of looping the components
wait for an event to happen, such as pressing the enter key. 
But of course it's the simplest way to write a parser.

## code in `src`

I see you make intensive use of the `use namespace std` which was
mentioned during the course as a bad practice. 

I liked the fact that you used exceptions, although I think they were
not part of the course. 

What I don't like is that you mainly used an imperative procedural 
approach where all is done via functions but most of what you did 
could have been implemented with objects and methods to interact
with these objects.

Your shared data source is a CSV file and you re-read over and over
the files. It would have made more sense in the light of this course
to create an object containing the data, being read only once,
and then all the rest of the code could pick what needed from that object.

In other words the only object in this code seems to be the external
CSV library.

I think @joachimckh can add his views on this.

# ROOT

The ROOT code is called directly in the C++ and it looks ok.
My only comment is that you should avoid overusing large data types for
small things. A temperature does not require doubles.

# Summary

Overall the project is good, even if we would have liked to see more 
object oriented programming than procedural.

I think you should have avoided parsing the command line with C++ interactively,
it's really not good with that - unless you use a specific library.
I think you did a good job parsing the input text but it's so painful that one loses the BASH autocomplete
features. For example this doesn't work:

```shell
Apptainer[pflorido@cosmos-pkf ~/git/MNXB11/TeamF]> ls ~/git/MNXB11/TeamF/datasets/smhi-opendata_1_93220_20231007_155708_Karlstad.csv
/home/pflorido/git/MNXB11/TeamF/datasets/smhi-opendata_1_93220_20231007_155708_Karlstad.csv
Apptainer[pflorido@cosmos-pkf ~/git/MNXB11/TeamF]> ./main 
Enter the filepath: ~/git/MNXB11/TeamF/datasets/smhi-opendata_1_93220_20231007_155708_Karlstad.csv
Invalid filepath!
```

I think is usually better to create binaries to which you pass paths
and solve the paths there instead of parsing interactively.

Also the choice of chosing which analysis is very nice, but a bit
old school. Rarely you will have that a researcher wants to spend 
time in an interactive shell, one would rather batch script everything
by just passing command line commands such as

```shell
./main dataset.csv mode
```

The program however works very good so it is overall a good job, and
good handling of closing the windows once the program is terminated.

# Action(s) to take

No action to take.
